Sources/System/IORing.swift

```diff
@@ -0,0 +1,657 @@
+    func to_iovec() -> iovec {
+        iovec(iov_base: baseAddress, iov_len: count)
+    }
+}
+
+// all pointers in this struct reference kernel-visible memory
+@usableFromInline struct SQRing: ~Copyable {
+    let kernelHead: UnsafePointer<Atomic<UInt32>>
+    let kernelTail: UnsafePointer<Atomic<UInt32>>
+    var userTail: UInt32
+
+    // from liburing: the kernel should never change these
+    // might change in the future with resizable rings?
+    let ringMask: UInt32
+    // let ringEntries: UInt32 - absorbed into array.count
+
+    // ring flags bitfield
+    // currently used by the kernel only in SQPOLL mode to indicate
+    // when the polling thread needs to be woken up
+    let flags: UnsafePointer<Atomic<UInt32>>
+
+    // ring array
+    // maps indexes between the actual ring and the submissionQueueEntries list,
+    // allowing the latter to be used as a kind of freelist with enough work?
+    // currently, just 1:1 mapping (0..<n)
+    let array: UnsafeMutableBufferPointer<UInt32>
+}
+
+struct CQRing: ~Copyable {
+    let kernelHead: UnsafePointer<Atomic<UInt32>>
+    let kernelTail: UnsafePointer<Atomic<UInt32>>
+
+    // TODO: determine if this is actually used
+    var userHead: UInt32
+
+    let ringMask: UInt32
+
+    let cqes: UnsafeBufferPointer<io_uring_cqe>
+}
+
+public struct IOResource<T> {
+    public typealias Resource = T
+    @usableFromInline let resource: T
+    @usableFromInline let index: Int
+
+    internal init(
+        resource: T,
+        index: Int
+    ) {
+        self.resource = resource
+        self.index = index
+    }
+}
+
+public typealias IORingFileSlot = IOResource<UInt32>
+public typealias IORingBuffer = IOResource<iovec>
+
+extension IORingFileSlot {
+    public var unsafeFileSlot: Int {
+        return index
+    }
+}
+extension IORingBuffer {
+    public var unsafeBuffer: UnsafeMutableRawBufferPointer {
+        return .init(start: resource.iov_base, count: resource.iov_len)
+    }
+}
+
+@inline(__always)
+internal func _writeRequest(
+    _ request: __owned RawIORequest, ring: inout SQRing,
+    submissionQueueEntries: UnsafeMutableBufferPointer<io_uring_sqe>
+)
+    -> Bool
+{
+    let entry = _blockingGetSubmissionEntry(
+        ring: &ring, submissionQueueEntries: submissionQueueEntries)
+    entry.pointee = request.rawValue
+    return true
+}
+
+@inline(__always)
+internal func _blockingGetSubmissionEntry(
+    ring: inout SQRing, submissionQueueEntries: UnsafeMutableBufferPointer<io_uring_sqe>
+) -> UnsafeMutablePointer<
+    io_uring_sqe
+> {
+    while true {
+        if let entry = _getSubmissionEntry(
+            ring: &ring,
+            submissionQueueEntries: submissionQueueEntries
+        ) {
+            return entry
+        }
+        // TODO: actually block here instead of spinning
+    }
+
+}
+
+//TODO: omitting signal mask for now
+//Tell the kernel that we've submitted requests and/or are waiting for completions
+internal func _enter(
```

Comment:
AFAICT the implementation here for flushing can run into the problem previously encountered here:

https://github.com/apple/swift-nio/blob/140be96c6ba178ce44989b2440e6e116078977e2/Sources/NIOPosix/LinuxUring.swift#L207

See https://github.com/axboe/liburing/issues/309 for background.

Reply:
oops I totally missed this comment earlier, will take a look!

Reply:
Not sure you saw my other comment about flushing either - just sharing some experience from the NIO prototype usage.